src/watch.rs

```diff
@@ -195,6 +201,60 @@ impl Watch {
     }
 }
 
+/// Lists the dirs and files in a directory, as two vectors.
+/// Given path must be a readable directory.
+fn list_dir(dir: &Path) -> Result<(Vec<PathBuf>, Vec<PathBuf>), std::io::Error> {
+    let mut dirs = vec![];
+    let mut files = vec![];
+    for entry in std::fs::read_dir(dir)? {
+        let entry = entry?;
+        if entry.file_type()?.is_dir() {
+            dirs.push(entry.path())
+        } else {
+            files.push(entry.path())
+        }
+    }
+    Ok((dirs, files))
+}
+
+/// List all children of the given path.
+/// Recurses into directories.
```

Comment:
... but avoids being recursive :) Is there a reason (bechmarked performance or something like that) that you went for full-blown imperative style here? I suspect that `walk_path_topo` could call itself, simplifying the code a bit.

Reply:
It felt a lot easier to do than going for a more functional approach (especially for topological sort).
Also it doesn’t depend on the call stack, which is usually the better approach for imperative languages (though I don’t think we would have directories that deep, but you never know. There’s always that *one* user).

Reply:
Here's an alternative, functional way to write the same function.

```rust
fn walk_path_topo(path: PathBuf) -> Result<Vec<PathBuf>, std::io::Error> { 
    let p = vec![path.clone()];
    if !path.is_dir() {    
        return Ok(p);
    }                                                                  
    let (dirs, files) = list_dir(&path)?;    
    Ok(p 
        .into_iter()
        .chain(files.into_iter())
        .chain(
            dirs.into_iter()
                .map(walk_path_topo)
                .collect::<Result<Vec<Vec<PathBuf>>, _>>()?
                .into_iter()
                .flatten(),
        )
        .collect())
}
```

Reply:
It’s self recursive though, so the stack might blow for edge cases (very deep paths). Amazingly the order is the exact same, but I have no idea about the performance of the iterator chain, whereas with the internally mutable code I can see what is going on (mutable appending to a deque and popping from the beginning). Not that matters mutch, since we are doing I/O at any step.

I think I’ll keep the internally mutating code, since it’s fairly straightforward.

Reply:
But it’s great to see a use of `chain`, I will have to look at how that is implemented internally.